app/Specification.hs

```diff
@@ -0,0 +1,570 @@
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE ViewPatterns               #-}
+
+module Main where
+
+import Control.Lens
+import Control.Monad (void)
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Either.Combinators (rightToMaybe)
+import Data.Monoid (Last (Last))
+import GHC.Generics
+import Wallet.Emulator (knownWallet)
+import qualified Data.Map as M
+import qualified Ledger
+import qualified Ledger.Ada as Ada
+import qualified Ledger.Constraints as Constraints
+import qualified Ledger.Scripts as Scripts
+import qualified Ledger.Typed.Scripts as TScripts
+import qualified Ledger.Value as Value
+import qualified Plutus.Contract as Contract
+import qualified Plutus.Contracts.Currency as Currency
+import qualified Plutus.Trace.Emulator as Trace
+import qualified PlutusTx
+import qualified PlutusTx.Eq as PEq
+import qualified PlutusTx.List as PList
+import qualified PlutusTx.Numeric as PNum
+import qualified PlutusTx.Ord as POrd
+import qualified PlutusTx.Trace as PTrace
+
+import Prelude
+
+data CDPDatum 
+  = ManagerDatum 
+    { getUsers  :: [Ledger.PubKeyHash] } 
+  | UserDatum 
+    { getPubKey :: Ledger.PubKeyHash
+    , getLocked :: Integer
+    , getMinted :: Integer }
+
+data EPT
+  = EPT 
+    { getAToken :: Value.AssetClass
+    , getUToken :: Value.AssetClass
+    , getAmount :: Integer } 
+  deriving (FromJSON, ToJSON, Generic)
+
+data CDPParams
+  = CDPParams 
+    { ogetAToken :: Value.AssetClass  
+    , ogetUToken :: Value.AssetClass }
+
+data CDP
+
+data CDPActions 
+  = MkOpen Ledger.PubKeyHash 
+  | MkDeposit Integer 
+  | MkWithdraw Integer 
+  | MkMint Integer
+  | MkBurn Integer
+
+PlutusTx.unstableMakeIsData ''CDPActions
+PlutusTx.unstableMakeIsData ''CDPDatum
+PlutusTx.unstableMakeIsData ''CDPParams
+PlutusTx.unstableMakeIsData ''EPT
+PlutusTx.makeLift ''CDPParams
+
+instance TScripts.ValidatorTypes CDP where
+  type DatumType CDP = CDPDatum
+  type RedeemerType CDP = CDPActions
+
+{-# INLINEABLE mkValidator #-}
+mkValidator :: CDPParams -> CDPDatum -> CDPActions -> Ledger.ScriptContext -> Bool
+mkValidator CDPParams{..} d act ctx = 
+  case act of 
+    MkOpen k ->
+      case d of
+        ManagerDatum l ->
+          PTrace.traceIfFalse "Inputs do not contain the real Manager" (inVal == nftVal) &&
+          PTrace.traceIfFalse "Wrong signature" (Ledger.txSignedBy info k) &&
+          PTrace.traceIfFalse "Invalid manager datum" (managerDatumList l k) &&
+          PTrace.traceIfFalse "Invalid user datum" (checkUserDatum k 0 0)  &&
+          PTrace.traceIfFalse "Invalid user output value" (checkUserValue atkVal) &&
+          PTrace.traceIfFalse "Invalid manager output value" (checkManagerValue nftVal)
+        _ -> PTrace.traceError "Invalid datum" 
```

Comment:
Trace chỉ là in ra lỗi thôi, em phải thêm Bool vào chứ, trong trường hợp này để là False, không cần trace.